pkg/kapis/devops/v1alpha4/handler.go

```diff
@@ -0,0 +1,116 @@
+package v1alpha4
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/emicklei/go-restful"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/selection"
+	"kubesphere.io/devops/pkg/api"
+	"kubesphere.io/devops/pkg/api/devops/v1alpha3"
+	"kubesphere.io/devops/pkg/api/devops/v1alpha4"
+	"kubesphere.io/devops/pkg/apiserver/query"
+	v1alpha32 "kubesphere.io/devops/pkg/models/resources/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strconv"
+)
+
+type apiHandlerOption struct {
+	client client.Client
+}
+
+// apiHandler contains functions to handle coming request and give a response.
+type apiHandler struct {
+	apiHandlerOption
+}
+
+// newAPIHandler creates an APIHandler.
+func newAPIHandler(o apiHandlerOption) *apiHandler {
+	return &apiHandler{o}
+}
+
+func (h *apiHandler) listPipelineRuns() restful.RouteFunction {
+	return func(request *restful.Request, response *restful.Response) {
+		nsName := request.PathParameter("namespace")
+		pipName := request.PathParameter("pipeline")
+		branchName := request.QueryParameter("branch")
+		backward, err := strconv.ParseBool(request.QueryParameter("backward"))
+		if err != nil {
+			// by default, we have to guarantee backward compatibility
+			backward = true
+		}
+
+		queryParam := query.ParseQueryParameter(request)
+
+		// validate the Pipeline
+		var pip v1alpha3.Pipeline
+		err = h.client.Get(context.Background(), client.ObjectKey{Namespace: nsName, Name: pipName}, &pip)
+		if err != nil {
+			api.HandleError(request, response, err)
+			return
+		}
+
+		// build label selector
+		labelSelector, err := buildLabelSelector(queryParam, pip.Name, branchName)
+		if err != nil {
+			api.HandleError(request, response, err)
+			return
+		}
+
+		var prs v1alpha4.PipelineRunList
+		// fetch PipelineRuns
+		if err := h.client.List(context.Background(), &prs,
+			client.InNamespace(pip.Namespace),
+			client.MatchingLabelsSelector{Selector: labelSelector}); err != nil {
+			api.HandleError(request, response, err)
+			return
+		}
+		var transformFunc = v1alpha32.NoTransformFunc()
+		if backward {
+			transformFunc = backwardTransform()
+		}
+
+		apiResult := v1alpha32.DefaultList(convertPipelineRunsToObject(prs.Items), queryParam, v1alpha32.DefaultCompare(), v1alpha32.DefaultFilter(), transformFunc)
+		_ = response.WriteAsJson(apiResult)
+	}
+}
+
+// backwardTransform transforms PipelineRun into JSON raw message of Jenkins run status.
+func backwardTransform() v1alpha32.TransformFunc {
+	return func(object runtime.Object) interface{} {
+		pr := object.(*v1alpha4.PipelineRun)
+		runStatusJSON := pr.Annotations[v1alpha4.JenkinsPipelineRunStatusKey]
+		rawRunStatus := json.RawMessage{}
+		// the error will never happen due to raw message won't be nil, so we ignore the error explicitly
+		_ = rawRunStatus.UnmarshalJSON([]byte(runStatusJSON))
```

Comment:
what will happen if the `runStatusJSON`was modified manually, and it turns out to an invalid json string.

Reply:
How about returning `{}` if the JSON is invalid?

Reply:
No matter what, cannot eat this error.

Reply:
Please see the first element below:

```json
{
 "items": [
  {},
  {
   "causes": [
    {
     "_class": "io.jenkins.blueocean.service.embedded.rest.AbstractRunImpl$BlueCauseImpl",
     "shortDescription": "Started by user admin",
     "userId": "admin",
     "userName": "admin"
    }
   ],
   "durationInMillis": 62234,
   "enQueueTime": "2021-09-08T21:57:55.633+08:00",
   "endTime": "2021-09-08T21:58:57.876+08:00",
   "estimatedDurationInMillis": 57378,
   "id": "5",
   "organization": "jenkins",
   "pipeline": "main",
   "replayable": true,
   "result": "SUCCESS",
   "runSummary": "stable",
   "startTime": "2021-09-08T21:57:55.642+08:00",
   "state": "FINISHED",
   "type": "WorkflowRun"
  },
  {
   "causes": [
    {
     "_class": "io.jenkins.blueocean.service.embedded.rest.AbstractRunImpl$BlueCauseImpl",
     "shortDescription": "Started by user admin",
     "userId": "admin",
     "userName": "admin"
    }
   ],
   "durationInMillis": 64198,
   "enQueueTime": "2021-09-08T14:22:14.762+08:00",
   "endTime": "2021-09-08T14:23:18.969+08:00",
   "estimatedDurationInMillis": 62034,
   "id": "4",
   "organization": "jenkins",
   "pipeline": "main",
   "replayable": true,
   "result": "SUCCESS",
   "runSummary": "stable",
   "startTime": "2021-09-08T14:22:14.771+08:00",
   "state": "FINISHED",
   "type": "WorkflowRun"
  }
 ],
 "totalItems": 3
}
```

Reply:
It's not reasonable. There are 3 items. But one of them is empty.

Reply:
Not necessary to return the message of invalid data to API response. Please use a logger to print it.

Reply:
Consider that there is no Jenkins run status yet, what data should be returned

Reply:
I prefer to return nothing. In this case, the total number should be 2.